### PR TITLE
bpo-1635741: port blake2 module to multi-phase init

### PR DESCRIPTION
**Pull request by koubaa**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by koubaa** [comment]
@vstinner @shihai1991  This is ready for review

File: Modules/_blake2/blake2b_impl.c [hunk 1]
```diff
@@ -391,47 +391,26 @@ py_blake2b_dealloc(PyObject *self)
         PyThread_free_lock(obj->lock);
         obj->lock = NULL;
     }
+
+    PyObject *type = Py_TYPE(self);
```
Comment (vstinner) [inline Modules/_blake2/blake2b_impl.c]:
```suggestion
    PyTypeObject *type = Py_TYPE(self);
```

Same comment in the other file.

File: Modules/_blake2/blake2module.c [hunk 2]
```diff
@@ -12,62 +12,82 @@
...
 PyDoc_STRVAR(blake2mod__doc__,
 "_blake2b provides BLAKE2b for hashlib\n"
 );
 
+typedef struct {
+    PyTypeObject* blake2b_type;
+    PyTypeObject* blake2s_type;
+} Blake2State;
+
+static inline Blake2State* blake2_get_state(PyObject *module)
```
Comment (vstinner) [inline Modules/_blake2/blake2module.c]: PEP 7 coding style:

```suggestion
static inline Blake2State*
blake2_get_state(PyObject *module)
```

File: Modules/_blake2/blake2module.c [hunk 3]
```diff
@@ -12,62 +12,82 @@
...
+{
+    _blake2_clear((PyObject *)module);
+}
 
 #define ADD_INT(d, name, value) do { \
     PyObject *x = PyLong_FromLong(value); \
     if (!x) { \
-        Py_DECREF(m); \
-        return NULL; \
+        Py_DECREF(x); \
```
Comment (vstinner) [inline Modules/_blake2/blake2module.c]: DECREF(NULL) is a bug.

File: Modules/_blake2/blake2module.c [hunk 4]
```diff
@@ -12,62 +12,82 @@
...
+        return -1; \
     } \
     Py_DECREF(x); \
 } while(0)
 
 
-PyMODINIT_FUNC
-PyInit__blake2(void)
+static int blake2_exec(PyObject *m)
```
Comment (vstinner) [inline Modules/_blake2/blake2module.c]: PEP 7:

```suggestion
static int
blake2_exec(PyObject *m)
```

File: Modules/_blake2/blake2s_impl.c [hunk 5]
```diff
@@ -391,47 +390,26 @@ py_blake2s_dealloc(PyObject *self)
...
-    0,                        /* tp_base            */
-    0,                        /* tp_dict            */
-    0,                        /* tp_descr_get       */
-    0,                        /* tp_descr_set       */
-    0,                        /* tp_dictoffset      */
-    0,                        /* tp_init            */
-    0,                        /* tp_alloc           */
-    py_blake2s_new,           /* tp_new             */
+// Using PyType_GetModuleState() on this type is safe since
+// it cannot be subclassed: it does not have the Py_TPFLAGS_BASETYPE flag.
```
Comment (vstinner) [inline Modules/_blake2/blake2s_impl.c]: You don't use PyType_GetModuleState(), so this comment can be removed.

Same comment in the other file.

File: Modules/_blake2/blake2module.c (hunk 2 above)
Comment (koubaa) [reply Modules/_blake2/blake2module.c]: @vstinner hopefully the [pep7 linter issue](https://github.com/giampaolo/psutil/issues/1351) is fixed so you don't have to spend time on this..

File: Misc/NEWS.d/next/Core and Builtins/2020-08-13-07-18-05.bpo-1635741.FC13e7.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Port :mod:`_blake2` to multi-phase init
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2020-08-13-07-18-05.bpo-1635741.FC13e7.rst]:
```suggestion
Port the :mod:`_blake2` extension module to the multi-phase initialization API (:pep:`489`).
```

**Review by vstinner — APPROVED** [review]
LGTM, with a minor suggestion on the NEWS entry (please update it).

But right now, I cannot merge the PR: https://mail.python.org/archives/list/python-dev@python.org/thread/KMXK6BQWQZRR2TGVRLECMR7CUVIVTUMQ/

Please ping me when the workflow will be repaired, so I will merge your PR.

**Comment by koubaa** [comment]
@vstinner looks like the doc issue is fixed in the CI.

**Comment by vstinner** [comment]
Merged, thanks @koubaa.

**Pull request merged**